modules/core/client/utils/validation.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * @typedef {array} rule - first element is testing function, second element is error
+ * testing function has value as the first argument
+ *    and all values as the second optional argument
+ * error can be anything
+ */
+/**
+ * provided a dictionary of rules and a dictionary of values,
+ * return a dictionary of errors for each value
+ *
+ * @param {Object.<string, rule[]>} ruleDict - dictionary of array of rules
+ * @param {Object.<string, any>} valueDict - dictionary of values
+ * @returns {Object.<string, any[]>} - dictionary of array of errors
+ */
+export const validate = (ruleDict, valueDict) => {
```

Comment:
I can imagine we might have much more extensive use for declarative clientside validation at some point, and it might be worth using a third party library for it.

Considered in that scope, then this seems fine, otherwise I might propose a structure that lets us define the rules seperate to the validation function, something like:

```js
const validate = createValidator(ruleDict);
const errors = validate(valueDict);
```

Reply:
2eb41973dc95f1b9dc6fd600f797f8f9b6d6669f
Made `createValidator` function. I wouldn't like to decide about the final version of client-side validation now. Let's leave it undecided until we're more sure what we need.